<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Instructions" data-toc-modified-id="Instructions-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Instructions</a></span></li><li><span><a href="#Before-You-Begin" data-toc-modified-id="Before-You-Begin-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Before You Begin</a></span></li><li><span><a href="#Imports" data-toc-modified-id="Imports-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Imports</a></span></li><li><span><a href="#Questions" data-toc-modified-id="Questions-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Questions</a></span></li><li><span><a href="#Food-Production---Formulation-Only" data-toc-modified-id="Food-Production---Formulation-Only-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Food Production - Formulation Only</a></span><ul class="toc-item"><li><span><a href="#Problem-Formulation" data-toc-modified-id="Problem-Formulation-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Problem Formulation</a></span><ul class="toc-item"><li><span><a href="#Decision-Variables" data-toc-modified-id="Decision-Variables-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Decision Variables</a></span></li><li><span><a href="#Objective" data-toc-modified-id="Objective-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Objective</a></span></li><li><span><a href="#Constraints" data-toc-modified-id="Constraints-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Constraints</a></span></li><li><span><a href="#Solution" data-toc-modified-id="Solution-5.1.4"><span class="toc-item-num">5.1.4&nbsp;&nbsp;</span>Solution</a></span></li></ul></li></ul></li><li><span><a href="#Production-Planning" data-toc-modified-id="Production-Planning-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Production Planning</a></span><ul class="toc-item"><li><span><a href="#Problem-Formulation" data-toc-modified-id="Problem-Formulation-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Problem Formulation</a></span><ul class="toc-item"><li><span><a href="#Decision-Variables" data-toc-modified-id="Decision-Variables-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Decision Variables</a></span></li><li><span><a href="#Objective-Function" data-toc-modified-id="Objective-Function-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>Objective Function</a></span></li><li><span><a href="#Constraint-Functions" data-toc-modified-id="Constraint-Functions-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>Constraint Functions</a></span></li><li><span><a href="#Solution" data-toc-modified-id="Solution-6.1.4"><span class="toc-item-num">6.1.4&nbsp;&nbsp;</span>Solution</a></span></li></ul></li></ul></li><li><span><a href="#Supply-Chain-Planning" data-toc-modified-id="Supply-Chain-Planning-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Supply Chain Planning</a></span><ul class="toc-item"><li><span><a href="#Problem-Formulation" data-toc-modified-id="Problem-Formulation-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Problem Formulation</a></span><ul class="toc-item"><li><span><a href="#Objective-Function" data-toc-modified-id="Objective-Function-7.1.1"><span class="toc-item-num">7.1.1&nbsp;&nbsp;</span>Objective Function</a></span></li><li><span><a href="#Constraint-Functions" data-toc-modified-id="Constraint-Functions-7.1.2"><span class="toc-item-num">7.1.2&nbsp;&nbsp;</span>Constraint Functions</a></span></li><li><span><a href="#Solution" data-toc-modified-id="Solution-7.1.3"><span class="toc-item-num">7.1.3&nbsp;&nbsp;</span>Solution</a></span></li></ul></li></ul></li></ul></div>

<h1>Instructions</h1>
<br>
<font size="+1">
    <ul>
        <li style="color:red">This problem set assesses your ability to analyze a given business decision by creating a concrete formulation of an optimization model, as well as practice solving your optimizaton model with CVXPY and summarizing the results. You should typeset math within Markdown using $\LaTeX$.</li>
        <br>
        <li><b>Recommended Steps</b></li>
        <br>
        <ol>
            <li>Read the prompt and summarize the task in your own words, using as few words as possible for ease of comprehension.</li>
            <br>
            <li>Pretend to be the computer program that you are trying to create. Examine the sample inputs and manually produce the appropriate results. Afterward, verify the correctness of your logic by comparing with the given sample outputs.</li>
            <br>
            <li>On a scrap piece of paper, write step-by-step instructions so that someone else can pretend to be the computer program and produce the appropriate results for any possible inputs. These instructions should target a human audience who does not know the problem description and only has your instructions as a guide. These instructions should serve as the roadmap for your code.</li>
            <br>
            <li>On another scrap piece of paper, write fragments of Python code to implement each step in your roadmap, focusing especially on the trickiest parts of the logic.</li>
            <br>
            <li>With your roadmap and code fragments from the previous steps, type Python code into a new Jupyter notebook cel to solve the given problem. You should type incrementally and run the code whenever you add a small chunk. This allows you to correct errors as they occur. It may also be helpful to print intermediate results to verify that the code is carrying out the logic as desired.</li>
            <br>
            <li>Run your code with some sample inputs and verify that the outputs are exactly the same as the sample outputs.</li>
            <br>
            <li>Once all of your problems are finished, it is useful to click <i>Kernel</i> $\rightarrow$ <i>Restart & Run All</i> to verify that your outputs for all of your problems continue to be correct. This helps to ensure that someone else will be able to replicate your results when running your notebook.</li>
            <br>
            <li style="color:red">Submit the <i>.ipynb</i> file on TurnItIn on Blackboard by the deadline. Be sure to refresh the page and double check your submission actually went through. <b>Note that you only need to submit your solutions, not all of the other recommended steps.</b> The recommended steps are meant to serve as a guide for your thinking process.</li>
            <br>
        </ol>
        <br>
        <hr style="border: 10px solid black">
        <br>
        <li><b>Grading</b></li>
        <br>
        <li>There are four possible scores you can get from submitting this assignment on time (submitting a blank file or one without any apparent effort does not count). Note that the rubric is designed to incentivize you to go for $100%$ mastery of the material, as the little details matter in programming.</li>
        <br>
        <ul>
            <li>Grade of $5$ out of $5$ - perfect submission with no significant errors</li>
            <br>
            <li>Grade of $4$ out of $5$ - near perfect submission with one or more significant errors</li>
            <br>
            <li>Grade of $2$ out of $5$ - apparent effort but far from perfect</li>
            <br>
            <li>Grade of $0$ out of $5$ - no submission or no apparent effort</li>
            <br>
        </ul>
    </ul>
</font>

<hr style="border: 20px solid black">

<h1>Before You Begin</h1>
<br>
<font size="+1">
    <ul>
        <li>Please read: <b>01_Practice_with_CVXPY_Optimization_and_Operations_Management.ipynb</b></li>
        <br>
        <li>Please read: <b>02_Optimal_Network_Design_For_Supply_Chains.ipynb</b></li>
        <br>
    </ul>
</font>

<hr style="border: 20px solid black">

<h1>Imports</h1>
<br>
<font size="+1">
    <ul>
        <li>Put your imports below.</li>
        <br>
        <li>Import Numpy and Pandas and CVXPY</li>
        <br>
    </ul>
</font>

In [2]:
import cvxpy as cp
import numpy as np
import pandas as pd

<hr style="border: 20px solid black">

<h1>Questions</h1>

<h1>Food Production - Formulation Only</h1>
<br>
<font size="+1">
    <ul>
        <li>A food factory requires $2000$ tons of canola oil every month as a raw ingredient.</li>
        <br>
        <li>The prices of canola oil fluctuates from month to month due to market conditions.</li>
        <br>
        <li>The predicted prices for the next six months are: <br> <table>
  <tr>
    <th>Month</th>
    <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
  </tr>
  <tr>
    <td>Price per ton</td>
    <th>150</th>
      <th>160</th>
      <th>180</th>
      <th>170</th>
      <th>180</th>
      <th>160</th>
  </tr>
    </table></li>
        <br>
        <li>The factory's supplier for canola oil delivers it on the first day of every month, and charges the above prices.</li>
        <br>
        <li>The factory can decide how much oil to buy each month from the supplier.</li>
        <br>
        <li>At the end of each month, the factory can also store unused oil for future use, but the inventory of oil (total amount stored)  cannot exceed $1000$ tons at any given time. Assume the inventory before the first shipment is zero.</li>
        <br>
        <li style="color:red">Formulate, in English and symbolically, an optimization problem to decide how much canola oil to buy for each of the six months in order to minimize the total purchase cost over the next six months.</li>
        <br>
        <li style="color:red">You do not have to use CVXPY to estimate the optimal purchasing policy and the associated purchasing cost, though you may choose to if interested, you only need to <b>formulate</b> the problem.</li>
        <br>
    </ul>
</font>

<h2>Problem Formulation</h2>
<br>
<font size="+1">
    <ul>
        <li>This consists of determining the decision variables, objective function, and constraint functions.</li>
        <br>
    </ul>
</font>

<h3>Decision Variables</h3>
<br>
<font size="+1">
    <ul>
        <li>What are the decision variables?</li>
        <br>
    </ul>
</font>

<br>
<font size="+1">
    <ul>
        <li> Decision Variable B: the number tons of canola oil to buy every month</li>
        <br>
    </ul>
</font>


In [3]:
B = cp.Variable(6, integer = True)

<h3>Objective</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the objective and the associated objective function?</li>
        <br>
    </ul>
</font>

$$\begin{align}
\text{ We want to minimize: } & 150B_1 +160B_2 + 180B_3 + 170B_4 + 180B_5 + 160B_6 & \\ \\
\end{align}$$

In [4]:
obj = cp.Minimize(150*B[0] +160*B[1] + 180*B[2] + 170*B[3] + 180*B[4] + 160*B[5])

<h3>Constraints</h3>
<br>
<font size="+1">
    <ul>
        <li>What are the constraints imposed by the factory?</li>
        <br>
    </ul>
</font>

<br>
<font size="+1">
    <ul>
        <li> Constraints:  </li>
         <li> 1. The number of tons of oil available for production each month must be at least 2,000 tons  </li>
        <li> 2. The number of tons of inventory at any month must be lower than or equal to 1,000 tons  </li>
        <li> 3. Also, we should notice that inventory must be non-negative </li>
        <br>
    </ul>
</font>

$$\begin{align}
\text{ minimize: } & 150B_1 +160B_2 + 180B_3 + 170B_4 + 180B_5 + 160B_6 & \\ \\
\text{ subject to: } & \\
& \text{ tons of oil available at month 1: } B_1 & \geq 2,000 \\
& \text{ tons of oil available at month 2: } B_2 +  (B_1 - 2,000) & \geq 2,000 \\
& \text{ tons of oil available at month 3: } B_3 + (B_1 + B_2 - 4,000)  & \geq 2,000 \\
& \text{ tons of oil available at month 4: } B_4  + (B_1 + B_2 + B_3 - 6,000)&\geq 2,000 \\
& \text{ tons of oil available at month 5: } B_5  + (B_1 + B_2 + B_3 + B_4 - 8,000) & \geq 2,000 \\
& \text{ tons of oil available at month 6: } B_6  +  (B_1 + B_2 + B_3 + B_4 + B_5 - 10,000) &\geq 2,000 \\
& \text{ inventory at the end of month 1: } (B_1 - 2,000) & \leq 1,000 \\
& \text{ inventory at the end of month 2: } (B_1 + B_2 - 4,000) & \leq 1,000 \\
& \text{ inventory at the end of month 3: } (B_1 + B_2 + B_3 - 6,000) & \leq 1,000 \\
& \text{ inventory at the end of month 4: }  (B_1 + B_2 + B_3 + B_4 - 8,000) & \leq 1,000 \\
& \text{ inventory at the end of month 5: }  (B_1 + B_2 + B_3 + B_4 + B_5 - 10,000) & \leq 1,000 \\
& \text{ inventory at the end of month 6: }  (B_1 + B_2 + B_3 + B_4 + B_5 + B_6 - 12,000) & \leq 1,000 \\
\end{align}$$


In [5]:
constraint_1 = (B[0] >= 2000)

constraint_2 = (B[1] + (B[0] -2000) >= 2000)

constraint_3 = (B[2] + (B[0] + B[1] -4000) >= 2000)

constraint_4 = (B[3] + (B[0] + B[1] + B[2] -6000) >= 2000)

constraint_5 = (B[4] + (B[0] + B[1] + B[2] + B[3] -8000) >= 2000)

constraint_6 = (B[5] +(B[0] + B[1] + B[2] + B[3] + B[4] -10000) >= 2000)

constraint_7 = (B[0] -2000) <= 1000

constraint_8 =  (B[0] + B[1] -4000) <= 1000

constraint_9 = (B[0] + B[1] + B[2] -6000) <= 1000

constraint_10 = (B[0] + B[1] + B[2] + B[3] -8000) <= 1000

constraint_11 = (B[0] + B[1] + B[2] + B[3] + B[4] -10000) <=1000

constraint_12 = (B[0] + B[1] + B[2] + B[3] + B[4] + B[5] -12000) <=1000

constraint_13 = (B>=0)

constraint_14 = (B[0] -2000) >=0

constraint_15 =  (B[0] + B[1] -4000) >=0

constraint_16 = (B[0] + B[1] + B[2] -6000) >=0

constraint_17 = (B[0] + B[1] + B[2] + B[3] -8000) >=0

constraint_18 = (B[0] + B[1] + B[2] + B[3] + B[4] -10000) >=0

constraint_19 = (B[0] + B[1] + B[2] + B[3] + B[4] + B[5] -12000) >=0


<h3>Solution</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the solution to the problem?</li>
        <br>
        <li>How can you interpret the solution?</li>
        <br>
        <li style="color:red">This step isn't required for this problem.</li>
        <br>
    </ul>
</font>

In [6]:
prob = cp.Problem(obj, [constraint_1, constraint_2, constraint_3,
                       constraint_4, constraint_5, constraint_6,
                       constraint_7, constraint_8, constraint_9,
                       constraint_10, constraint_11, constraint_12,
                       constraint_13, constraint_14, constraint_15,
                       constraint_16, constraint_17, constraint_18,
                       constraint_19])

In [7]:
prob.solve()

1960000.0

In [8]:
B.value

array([3000., 2000., 1000., 3000., 1000., 2000.])

<br>
<font size="+1">
    <ul>
        <li>This tells us the food factory should buy 3000 tons, 2000 tons, 1000 tons, 3000 tons, 1000 tons, and 2000 tons of canola oil in the following six months respectively, which gives a minimum purchasing cost of $1,960,000.</li>
        <br>
    </ul>
</font>

<hr style="border: 20px solid black">

<h1>Production Planning</h1>
<br>
<font size="+1">
    <ul>
        <li>The Magnetron Company manufactures and markets microwave ovens.</li>
        <br>
        <ul>
            <li>Currently, the company produces two models: full-size and compact.</li>
            <br>
            <li>Production is limited by the amount of labor available in the general assembly and electronics assembly departments, as well as by the demand for each model.</li>
            <br>
            <li>Each full-size oven requires $2$ hours of general assembly and $2$ hours of electronic assembly, whereas each compact oven requires $2$ hours of general assembly and $3$ hours of electronic assembly.</li>
            <br>
            <li>In the current production period, there are $500$ hours of general assembly labor available and $800$ hours of electronic assembly labor available.</li>
            <br>
            <li>Additionally, the company estimates it can sell at most $220$ full-size ovens and $180$ compact ovens in the current production period.</li>
            <br>
            <li>The earnings contribution per oven is $120$ dollars for a full-size oven and $130$ dollars for a compact oven.</li>
            <br>
            <li style="color:red">The company would like to find an earnings-maximizing production plan for the current production period.</li>
            <br>
        </ul>
        <li style="color:blue">For this problem, formulate the problem in English, as well as in equations. Then, solve the problem using CVXPY by finding the maximum earnings and the associated production plan that yields the maximum earnings.</li>
        <br>
    </ul>
</font>

<h2>Problem Formulation</h2>
<br>
<font size="+1">
    <ul>
        <li>This consists of determining the decision variables, objective function, and constraint functions.</li>
        <br>
    </ul>
</font>

<h3>Decision Variables</h3>
<br>
<font size="+1">
    <ul>
        <li>What decisions does the company have to make, i.e. what is in their control?</li>
        <br>
        <ul>
            <li>You can assume the variables are continuous and not integer valued.</li>
            <br>
        </ul>
    </ul>
</font>

<br>
<font size="+1">
    <ul>
        <li>There are 2 decision variables:  </li>
         <li> 1. F: the number of full-size ovens to produce (which should be a non-negative)  </li>
        <li> 2. C: the number of compact ovens to product (which should be a non-negative) </li>
        <br>
    </ul>
</font>

In [9]:
F = cp.Variable(1, nonneg = True)

C = cp.Variable(1, nonneg = True)

<h3>Objective Function</h3>
<br>
<font size="+1">
    <ul>
        <li>What's the goal?</li>
        <br>
    </ul>
</font>

<br>
<font size="+1">
    <ul>
        <li>$$\begin{align}
\text{ We want to maximize the earnings: } & 120 \cdot F + 130 \cdot C & \\ \\
\end{align}$$ </li>
        <br>
    </ul>
</font>

In [10]:
obj = cp.Maximize(120*F + 130*C)

<h3>Constraint Functions</h3>
<br>
<font size="+1">
    <ul>
        <li>Recall, you want to find the production plan (i.e. number of units of full-size and compact microwave ovens produced) that maximizes earnings while satisfying assembly and demand constraints.</li>
        <br>
    </ul>
</font>

<br>
<font size="+1">
    <ul>
        <li>Constraints:  </li>
         <li> 1. There are 500 hours of general assembly labor available, and 800  hours of electronic assembly labor available  </li>
        <li> 2. The companyit can sell at most 220 full-size ovens and 180 compact ovens </li>
        <br>
    </ul>
</font>

In [11]:
constraint_1 = (2*F + 2*C <= 500)

constraint_2 = (2*F + 3*C <=800)

constraint_3 = (F <= 220)

constraint_4 = (C <= 180)

<h3>Solution</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the solution to the problem?</li>
        <br>
        <li>How can you interpret the solution?</li>
        <br>
    </ul>
</font>

$$\begin{align}
\text{ maximize: } & 120 \cdot F + 130 \cdot C & \\ \\
\text{ subject to: } & 2 \cdot F + 2 \cdot C & \leq 500 \\
& 2 \cdot F + 3 \cdot C & \leq 800 \\
& F & \leq 220 \\
& C & \leq 180 \\
\end{align}$$

In [12]:
prob = cp.Problem(obj, [constraint_1, constraint_2, constraint_3, constraint_4])

In [13]:
prob.solve()

31799.99999951865

In [14]:
F.value

array([70.])

In [15]:
C.value

array([179.99999999])

<br>
<font size="+1">
    <ul>
        <li>This tells us the production plan is to produce $70$ units of the full-size ovens and $180$ units of the compact ovens, which gives a maximum earnings of $31,800$ dollars per this production period.</li>
        <br>
    </ul>
</font>

<hr style="border: 20px solid black">

<h1>Supply Chain Planning</h1>
<br>
<font size="+1">
    <ul>
        <li style="color:red">Warning: This problem is challenging and will not be graded by correctness, but rather effort, give it your best shot.</li>
        <br>
        <li>Every time Amazon has to ship an item from its fulfillment centers to a particular region, it incurs a shipping cost.</li>
        <br>
        <li>Consider the following shipping costs from three of Amazon's fulfillment centers (FC) to four regions (A, B, C, and D):</li>
        <br>
        <li><br><table>
  <tr>
    <th>Region \ FC:</th>
    <th>1</th>
    <th>2</th>
      <th>3</th>
  </tr>
  <tr>
    <td>A: King's County, NY</td>
    <td>20</td>
    <td>8</td>
      <td>25</td>
  </tr>
            <tr>
    <td>B: LA County, CA</td>
    <td>18</td>
      <td>23</td>
      <td>8</td>
  </tr>
  <tr>
    <td>C: King County, WA</td>
    <td>21</td>
    <td>24</td>
      <td>8</td>
  </tr>
            <tr>
    <td>D: Harris County, TX</td>
    <td>8</td>
    <td>8</td>
      <td>19</td>
  </tr>
       </table></li>
        <br>
        <li>It costs $20$ dollars to ship from fulfillment center $1$ to region $A$, for example.</li>
        <br>
        <li>We also have weekly demand estimates for the item in each region:</li>
        <br>
        <li>
        <table>
  <tr>
    <th>Region A</th>
    <th>Region B</th>
    <th>Region C</th>
      <th>Region D</th>
  </tr>
             <tr>
    <th>30</th>
    <th>50</th>
    <th>10</th>
      <th>20</th>
            </tr></table></li>
        <br>
        <li>Further, assume each fulfillment center is able to ship up to $40$ units each week in total.</li>
        <br>
        <li>What is the minimum transportation cost needed to satisfy all weekly demand while respecting fulfillment center capacities, as well as the optimal shipment plan?</li>
        <br>
        <li style="color:red">In other words, what is the shipment plan (i.e. number of units from what FC to what region) that minimizes transportation costs while satisfying total demand and not going above fulfillment center capacities?</li>
        <br>
        <li style="color:blue">For this problem, formulate the problem in English, as well as in equations. Then, solve the problem using CVXPY by finding the minimum costs and the associated shipment plan that yields the minimum costs.</li>
        <br>
    </ul>
</font>

<h2>Problem Formulation</h2>
<br>
<font size="+1">
    <ul>
        <li>This consists of determining the decision variables, objective function, and constraint functions.</li>
        <br>
    </ul>
</font>

<br>
<font size="+1">
    <ul>
        <li>What decisions do you, the transportation designer, have to make, i.e. what is in your control?</li>
        <br>
        <li>You have to decide how many units to transport from each fulfillment center to each region.</li>
        <br>
        <ul>
            <li>$x_{ij} = $ number of units to tranport from FC $i \in \{1,2,3\}$ to region $j \in \{A, B, C, D\}$.</li>
            <br>
            <ul>
                <li>note this is continuous variable </li>
                <br>
            </ul>
        </ul>
        <li>Now, we can instantiate a CVXPY variable object to represent the decision variables.</li>
        <br>
    </ul>
</font>

In [16]:
flow = cp.Variable((3,4), nonneg=True)

<h3>Objective Function</h3>
<br>
<font size="+1">
    <ul>
        <li>What's the goal?</li>
        <br>
    </ul>
</font>

<br>
<font size="+1">
    <ul>
        <li>The goal is to minimize the total shipping costs: $$\text{minimize: } C = \sum_{(i,j) \in \text{FC} \times \text{Region}}{\text{cost}_{i,j}\times \text{flow}_{i,j}}$$</li>
        <br>
    </ul>
</font>

In [17]:
costs = pd.DataFrame({
    ('1', 'A'): 20,
    ('2', 'A'): 8,
    ('3', 'A'): 25,
    ('1', 'B'): 18,
    ('2', 'B'): 23,
    ('3', 'B'): 8,
    ('1', 'C'): 21,
    ('2', 'C'): 24,
    ('3', 'C'): 8,
    ('1', 'D'): 8,
    ('2', 'D'): 8,
    ('3', 'D'): 19
}, index=['costs']).T

costs

costs
1 A     20
2 A      8
3 A     25
1 B     18
2 B     23
3 B      8
1 C     21
2 C     24
3 C      8
1 D      8
2 D      8
3 D     19

In [18]:
costs = costs.unstack()

costs

costs            
      A   B   C   D
1    20  18  21   8
2     8  23  24   8
3    25   8   8  19

In [19]:
total_flow_cost = sum([costs.values[i,j]*flow[i,j] for i in range(costs.shape[0]) for j in range(costs.shape[1])])

In [20]:
objective_function = cp.Minimize(total_flow_cost)

<h3>Constraint Functions</h3>
<br>
<font size="+1">
    <ul>
        <li>Recall, you want to find the shipment plan (i.e. number of units from a FC to a region) that minimizes transportation costs while satisfying total demand and not going above fulfillment center capacities.</li>
        <br>
    </ul>
</font>

In [22]:
# Create data frame to capture fulfillment center limits

FC = pd.DataFrame({'1': 40,
                    '2': 40,
                    '3': 40,}, index=['FC_limits']).T

FC

FC_limits
1         40
2         40
3         40

In [23]:
# Create data frame to caputure customer demand

demand = pd.DataFrame({'A': 30,
               'B': 50,
               'C': 10,
               'D': 20,}, index=['demand']).T

demand

demand
A      30
B      50
C      10
D      20

In [24]:
FC_output_constraints = [sum([flow[0, j] for j in [0,1,2,3]]) <= 40, 
                              sum([flow[1, j] for j in [0,1,2,3]]) <= 40,
                          sum([flow[2, j] for j in [0,1,2,3]]) <= 40
                        ]

FC_output_constraints

[Inequality(Expression(AFFINE, NONNEGATIVE, ())),
 Inequality(Expression(AFFINE, NONNEGATIVE, ())),
 Inequality(Expression(AFFINE, NONNEGATIVE, ()))]

In [25]:
customer_demand_constraints = [sum([flow[i, 0] for i in [0,1,2]]) == demand.loc['A', 'demand'],
                              sum([flow[i, 1] for i in [0,1,2]]) == demand.loc['B', 'demand'],
                              sum([flow[i, 2] for i in [0,1,2]]) == demand.loc['C', 'demand'],
                              sum([flow[i, 3] for i in [0,1,2]]) == demand.loc['D', 'demand']
                              ]

customer_demand_constraints

[Equality(Expression(AFFINE, NONNEGATIVE, ()), Constant(CONSTANT, NONNEGATIVE, ())),
 Equality(Expression(AFFINE, NONNEGATIVE, ()), Constant(CONSTANT, NONNEGATIVE, ())),
 Equality(Expression(AFFINE, NONNEGATIVE, ()), Constant(CONSTANT, NONNEGATIVE, ())),
 Equality(Expression(AFFINE, NONNEGATIVE, ()), Constant(CONSTANT, NONNEGATIVE, ()))]

<h3>Solution</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the solution to the problem?</li>
        <br>
        <li>How can you interpret the solution?</li>
        <br>
    </ul>
</font>

In [26]:
constraints = FC_output_constraints + customer_demand_constraints

In [27]:
problem = cp.Problem(objective_function, constraints)

In [28]:
problem.solve(feastol=1e-54, solver='ECOS_BB')

C:\Users\DELL\Anaconda3\lib\site-packages\cvxpy\problems\problem.py:1338: UserWarning: Solution may be inaccurate. Try another solver, adjusting the solver settings, or solve with verbose=True for more information.
  "Solution may be inaccurate. Try another solver, "


1080.000000000002

In [29]:
np.round(flow.value)

array([[ 0., 20.,  0., 14.],
       [30.,  0.,  0.,  6.],
       [ 0., 30., 10.,  0.]])

In [30]:
# Weekly shipping plan x_{i,j} to ship from fulfillment centers (i) to regions (j)
# stored in the matrix below

[[ 0. , 20. ,  0. , 13.8],
[30. ,  0. ,  0. ,  6.2],
[0. , 30. , 10. , 0. ]]

[[0.0, 20.0, 0.0, 13.8], [30.0, 0.0, 0.0, 6.2], [0.0, 30.0, 10.0, 0.0]]

<br>
<font size="+1">
    <ul>
        <li>This tells us the optimal shipment plan from fulfillment centers to regions that minimizes shipping costs to $1080$ dollars per week.</li>
        <br>
    </ul>
</font>

<hr style="border: 20px solid black">